HHH-12993 - Omit joining of superclass table when querying subclass only

ladislavovic opened the pull request:
* it go through AST and find out which tables are referenced. If super
  class table in "joined" inheritance strategy is not referenced, is is
  not joined
* when any filter is active then all super class table are joined, because
  filter can reference them and I can not find it out from query AST analyze.
  It could be optimized probably somehow.

ladislavovic commented:
Damn, some tests failed. It depends on environment. Here 5 tests failed, on my Ubuntu 18_04 with openjdk-8 3 failed and on my dev machine with win10 and oracle-javva-8 are all tests OK. So I am solving that.

beikov commented on hibernate-core/src/main/java/org/hibernate/persister/entity/AbstractEntityPersister.java:
```diff
@@ -521,6 +521,8 @@ public boolean hasRowId() {
 		return propertySelectable;
 	}
 
+	public abstract String[] getTableNames();
```
Not sure if `AbstractEntityPersister` can be considered an API/SPI but IMO a default implementation should just return null here to be safe.

ladislavovic replied:
Thanks for the implementation of getTableNames(), I've done it like so.

beikov commented on hibernate-core/src/main/java/org/hibernate/hql/internal/ast/util/JoinProcessor.java:
```diff
@@ -94,9 +103,73 @@ public static JoinType toHibernateJoinType(int astJoinType) {
 	public void processJoins(QueryNode query) {
 		final FromClause fromClause = query.getFromClause();
 
+		Set<String> queryReferencedTables = findQueryReferencedTables( query );
```
IMO this should be deactivatable by a property so that one can deactivate this feature in case of a bug.

ladislavovic replied:
I Added `AvailableSettings.OMIT_JOIN_OF_SUPERCLASS_TABLES` property. If it is false, the feature is deactivated. Default value is true.

beikov commented on hibernate-core/src/test/java/org/hibernate/test/hql/HQLTest.java:
```diff
@@ -241,10 +241,13 @@ public void testCollectionMemberDeclarations(){
 	}
 
 	@Test
+	@FailureExpected(
```
When introducing a configuration option, you can re-enable these tests.

JWGmeligMeyling replied:
Imho these tests should never be deactivated, I'd rather see them adjusted to the new PR.

ladislavovic replied:
Tests are re-enabled already.

ladislavovic commented on hibernate-core/src/main/java/org/hibernate/persister/entity/UnionSubclassEntityPersister.java:
```diff
@@ -222,6 +222,11 @@ public String getTableName() {
 
+	@Override
+	public String[] getTableNames() {
+		return new String[] {tableName};
```
I think you do not have to be afraid about `UnionSubclassEntityPersister`. It totally ignores `referencedTables` collection. Now, when the implementation of getTableNames() is in `AbstractEntityPersister`, `UnionSubclassEntityPersister` has no change in comparison with master branch.

Only `JoinedSubclassEntityPersister` use this information to optimize table joining.

ladislavovic replied:
> I am not only worried about this issue with respect to subclass tables, but also for example in the context of secondary tables. I assume that if you only select a property from a secondary table, this will be the only table joined? That would be slightly odd behaviour. Can you for example add a test of entities A and B (not sharing the same root class) that reference a secondary table C, and then show the produced SQL and output for SELECT a.somePropOnC FROM A a? I assume this would now return somePropOnC values for B too, which is obviously wrong.

I added test for that, see `OmitAncestorJoinWithSecondaryTableTest`.

ladislavovic commented on hibernate-core/src/main/java/org/hibernate/persister/entity/AbstractEntityPersister.java:
```diff
@@ -3930,9 +3948,15 @@ protected JoinType determineSubclassTableJoinType(
 			boolean includeSubclasses,
-			Set<String> treatAsDeclarations) {
+			Set<String> treatAsDeclarations,
+			Set<String> referencedTables) {
```
I added the methods.

ladislavovic commented on hibernate-core/src/main/java/org/hibernate/persister/entity/SingleTableEntityPersister.java:
```diff
@@ -486,6 +486,11 @@ public String getTableName() {
 
+	@Override
+	public String[] getTableNames() {
+		return qualifiedTableNames;
```
I've briefly checked the issue an maybe this approach could be used to solve that. Now the collection `referencedTables` is passed to procedure which joining parent tables in `JoinedSubclassEntityPersister`. It could be probably passed also to other places to optimize joining.

beikov commented on hibernate-core/src/main/java/org/hibernate/cfg/AvailableSettings.java:
```diff
@@ -1961,4 +1961,18 @@
+	/**
+	 * <p>
+	 * When you use {@link javax.persistence.InheritanceType#JOINED} strategy for inheritance mapping and query
+	 * a value from a entity, all superclass tables are joined in the query regardless you need them. With
```
Typo `value from a entity` should be `value from an entity`

beikov commented on hibernate-core/src/main/java/org/hibernate/cfg/AvailableSettings.java:
```diff
@@ -1961,4 +1961,18 @@
+	 * <p>
+	 * When you use {@link javax.persistence.InheritanceType#JOINED} strategy for inheritance mapping and query
+	 * a value from a entity, all superclass tables are joined in the query regardless you need them. With
+	 * this settings set to true only superclass tables which are really needed are joined.
```
Typo `this settings` should be `this setting`

beikov commented on hibernate-core/src/main/java/org/hibernate/hql/internal/ast/util/JoinProcessor.java:
```diff
@@ -48,6 +59,9 @@
 	private static final CoreMessageLogger LOG = CoreLogging.messageLogger( JoinProcessor.class );
 
+	// TODO where to place the default value?
+	public static final String DEFAULT_OMIT_JOIN_OF_SUPERCLASS_TABLES = Boolean.TRUE.toString();
```
Please add the configuration option to `org.hibernate.boot.spi.SessionFactoryOptions`. Take a look at how `isCollectionJoinSubqueryRewriteEnabled()` is implemented and initialized. There you can see how to set the default value.

beikov commented on hibernate-core/src/main/java/org/hibernate/hql/internal/ast/util/JoinProcessor.java:
```diff
@@ -94,9 +108,87 @@ public static JoinType toHibernateJoinType(int astJoinType) {
+
+	private Set<String> findQueryReferencedTables(QueryNode query) {
+		Map<String, Object> sessionFactoryProperties = walker.getSessionFactoryHelper().getFactory().getProperties();
+		Object enabled = sessionFactoryProperties.getOrDefault(
```
Use `walker.getSessionFactoryHelper().getFactory().getSessionFactoryOptions().isOmitJoinOfSuperclassTablesEnabled()`

beikov commented on hibernate-core/src/main/java/org/hibernate/hql/internal/ast/util/JoinProcessor.java:
```diff
@@ -94,9 +108,87 @@ public static JoinType toHibernateJoinType(int astJoinType) {
+		);
+		boolean disabled = !Boolean.parseBoolean( enabled.toString() );
+		if ( disabled ) {
+			LOG.debug( "Finding of query referenced tables is skipped because the feature is disabled by " +
```
Put a `LOG.isDebugEnabled()` around this.

beikov commented on hibernate-core/src/main/java/org/hibernate/hql/internal/ast/util/JoinProcessor.java:
```diff
@@ -94,9 +108,87 @@ public static JoinType toHibernateJoinType(int astJoinType) {
 	public void processJoins(QueryNode query) {
 		final FromClause fromClause = query.getFromClause();
 
+		Map<String, Object> properties = walker.getSessionFactoryHelper().getFactory().getProperties();
```
Remove this line

JWGmeligMeyling commented:
Thanks for addressing the feedback. I'll have a look at the produced SQL from the tests and test the feature in my own domain and get back to you.

sebersole commented:
Please make sure we have plenty of tests for this.  Implementing this in 6 should be much easier, but the implementation will be completely different.  We want to make sure they functionality still works

ladislavovic commented on hibernate-core/src/main/java/org/hibernate/boot/internal/SessionFactoryOptionsBuilder.java:
```diff
@@ -206,6 +207,7 @@
 	private final boolean procedureParameterNullPassingEnabled;
 	private final boolean collectionJoinSubqueryRewriteEnabled;
 	private boolean jdbcStyleParamsZeroBased;
+	private boolean omitJoinOfSuperclassTablesEnabled;
```
Please which commits do you want to squash? You mean all 13 commits?

beikov replied:
Yeah so that there are only 2 commits in the end, one containing only the test and another one containing only the implementation

ladislavovic replied:
final is added

gsmet commented on hibernate-core/src/main/java/org/hibernate/boot/internal/SessionFactoryOptionsBuilder.java:
```diff
@@ -350,6 +352,7 @@ public SessionFactoryOptionsBuilder(StandardServiceRegistry serviceRegistry, Boo
 				CONVENTIONAL_JAVA_CONSTANTS, BOOLEAN, true );
 		this.procedureParameterNullPassingEnabled = cfgService.getSetting( PROCEDURE_NULL_PARAM_PASSING, BOOLEAN, false );
 		this.collectionJoinSubqueryRewriteEnabled = cfgService.getSetting( COLLECTION_JOIN_SUBQUERY, BOOLEAN, true );
+		this.omitJoinOfSuperclassTablesEnabled = cfgService.getSetting( OMIT_JOIN_OF_SUPERCLASS_TABLES, BOOLEAN, true );
```
Am I the only one who wouldn't make it the default behavior but only a possible optimization knob for advanced users?

beikov replied:
How would we get broad testing if we wouldn't enable the optimization by default and allow it to be deactivated for further baking if necessary? I think 5.4 is a good candidate for enabling this by default.

JWGmeligMeyling replied:
I'm not sure whether it can be allowed. The produced query results may be different for unconstrained inheritance hierarchies (a row that exists in a subclass table but not in a superclass table, which was previously omitted by the join but is now present in result sets). No blocker for me personally, and I can hardly believe that this will be a problem for anyone, but we seem to consider this very edge case at other locations so we may have to keep it in mind.

gsmet commented on hibernate-core/src/main/java/org/hibernate/cfg/AvailableSettings.java:
```diff
@@ -1976,4 +1976,19 @@
+	 * The default value is true.
+	 * </p>
+	 *
+	 * @Since 5.4
```
`@since` without a capital.

gsmet commented on hibernate-core/src/main/java/org/hibernate/cfg/AvailableSettings.java:
```diff
@@ -1976,4 +1976,19 @@
+	 *
+	 * @Since 5.4
+	 */
+	String OMIT_JOIN_OF_SUPERCLASS_TABLES = "hibernate.query.omit_join_of_superclass_tables";
```
I'm not a native speaker but the option name sounds weird to my ear. Any native speaker with a better proposal?

I would have gone with something like `lenient_superclass_table_join` because you don't really always omit them, you are just considering only the useful ones.

@sebersole you're usually good with finding names. Any better proposal?

JWGmeligMeyling replied:
I don't really like the property name either. Your suggestion looks fine.

gsmet commented on hibernate-core/src/main/java/org/hibernate/hql/internal/ast/tree/DotNode.java:
```diff
@@ -74,6 +74,12 @@ public QueryException buildIllegalCollectionDereferenceException(String property
+	 * The identifier that is the name of the property. In comparison with {@link #propertyName}
+	 * it is always identical with identifier in the query, it is not changed during processing.
+	 */
+	private String originalPropertyName;
```
missing new line below this line.

ladislavovic replied:
fixed

gsmet commented on hibernate-core/src/main/java/org/hibernate/hql/internal/ast/tree/DotNode.java:
```diff
@@ -690,6 +697,25 @@ public Type getDataType() {
+				FromElement fromElement = fromReferenceNode.getFromElement();
+				if ( fromElement != null ) {
+					String table = fromElement.getPropertyTableName( getOriginalPropertyName() );
+					if (table != null) {
```
Missing padding. Looks like our Checkstyle rules are not correctly enforced.

ladislavovic replied:
Fixed. I must set automatic formatting in IntelliJIdea properly...

gsmet commented on hibernate-core/src/main/java/org/hibernate/hql/internal/ast/tree/FromElementType.java:
```diff
@@ -368,6 +365,15 @@ public QueryableCollection getQueryableCollection() {
 
+	public String getPropertyTableName(String propertyName) {
+		checkInitialized();
+		if (this.persister != null) {
```
Same here about Checkstyle. I suppose we are not strict on the padding rule.

ladislavovic replied:
fixed

gsmet commented on hibernate-core/src/main/java/org/hibernate/hql/internal/ast/tree/FromReferenceNode.java:
```diff
@@ -137,4 +137,14 @@ protected boolean isFromElementUpdateOrDeleteRoot(FromElement element) {
 	}
 
+	/**
+	 * Returns a table names which are referenced by this node. If the tables
```
s/Returns a table names/Returns table names/

ladislavovic replied:
fixed

gsmet commented on hibernate-core/src/main/java/org/hibernate/hql/internal/ast/util/ASTPrinter.java:
```diff
@@ -120,6 +114,7 @@ private void showAst(ArrayList<AST> parents, PrintWriter pw, AST ast) {
 		}
 
 		showNode( pw, ast );
+		showNodeProperties(parents, pw, ast);
```
Padding.

ladislavovic replied:
fixed

gsmet commented on hibernate-core/src/main/java/org/hibernate/hql/internal/ast/util/JoinProcessor.java:
```diff
@@ -94,9 +105,84 @@ public static JoinType toHibernateJoinType(int astJoinType) {
+		return result;
+	}
+
+	private <T extends AST> void findAllNodesCore(AST node, Class<T> clazz, List<T> found) {
```
Maybe `doFindAllNodes`? I don't think we use `Core` anywhere else.

ladislavovic replied:
fixed

gsmet commented on hibernate-core/src/main/java/org/hibernate/hql/internal/ast/util/JoinProcessor.java:
```diff
@@ -94,9 +105,84 @@ public static JoinType toHibernateJoinType(int astJoinType) {
+				.getFactory()
+				.getSessionFactoryOptions()
+				.isOmitJoinOfSuperclassTablesEnabled() ) {
+			if (LOG.isDebugEnabled()) {
```
padding.

ladislavovic replied:
fixed

gsmet commented on hibernate-core/src/main/java/org/hibernate/hql/internal/ast/util/JoinProcessor.java:
```diff
@@ -94,9 +105,84 @@ public static JoinType toHibernateJoinType(int astJoinType) {
+			return null;
+		}
+
+		if ( walker.getEnabledFilters() != null && !walker.getEnabledFilters().isEmpty() ) {
```
See `CollectionHelper.isEmpty()`.

ladislavovic replied:
fixed

gsmet commented on hibernate-core/src/main/java/org/hibernate/hql/internal/ast/util/JoinProcessor.java:
```diff
@@ -94,9 +105,84 @@ public static JoinType toHibernateJoinType(int astJoinType) {
+			String[] tables = node.getReferencedTables();
+			if ( tables != null ) {
+				Arrays.stream( tables )
+						.filter( Objects::nonNull )
```
Can't we filter the nulls before adding them?

ladislavovic replied:
Yes, `getReferencedTables()` should not returns any null elements in the array, null check is useless. Removed.

gsmet commented on hibernate-core/src/main/java/org/hibernate/hql/internal/ast/util/JoinProcessor.java:
```diff
@@ -94,9 +105,84 @@ public static JoinType toHibernateJoinType(int astJoinType) {
+					AbstractEntityPersister aep = (AbstractEntityPersister) entityPersister;
+					String[] tables = aep.getTableNames();
+					Arrays.stream( tables )
+							.filter( Objects::nonNull )
```
Do we really expect null here?

ladislavovic replied:
again, null check is useless, removed.

gsmet commented on hibernate-core/src/main/java/org/hibernate/persister/entity/AbstractEntityPersister.java:
```diff
@@ -521,6 +521,14 @@ public boolean hasRowId() {
+	public String[] getTableNames() {
+		String[] tableNames = new String[ getTableSpan() ];
+		for ( int i = 0; i < tableNames.length; i++ ) {
+			tableNames[ i ] = getTableName(i);
```
I don't think we usually pad the index but the parameter should be.

ladislavovic replied:
fixed

gsmet commented on hibernate-core/src/main/java/org/hibernate/hql/internal/ast/util/JoinProcessor.java:
```diff
@@ -94,9 +105,84 @@ public static JoinType toHibernateJoinType(int astJoinType) {
+				FromElement fromElement = (FromElement) element;
+				EntityPersister entityPersister = fromElement.getEntityPersister();
+				if ( entityPersister != null && entityPersister instanceof AbstractEntityPersister ) {
+					AbstractEntityPersister aep = (AbstractEntityPersister) entityPersister;
```
Maybe it's a stupid question but what if `canOmitSuperclassTableJoin()` of this persister returns false?

ladislavovic replied:
It does not matter. This method looks for all tables which are referenced by the query. And persister here just returns all tables which are used for entity mapping because all these tables are referenced.

`canOmitSuperclassTableJoin()` is used later on when a persister creates SQL from part of the query. Then a persister, if the `canOmitSuperclassTableJoin()` is `true`, does not have to use all tables bud can omit some of them.

gsmet commented on hibernate-core/src/main/java/org/hibernate/hql/internal/ast/util/JoinProcessor.java:
```diff
@@ -94,9 +105,84 @@ public static JoinType toHibernateJoinType(int astJoinType) {
+				EntityPersister entityPersister = fromElement.getEntityPersister();
+				if ( entityPersister != null && entityPersister instanceof AbstractEntityPersister ) {
+					AbstractEntityPersister aep = (AbstractEntityPersister) entityPersister;
+					String[] tables = aep.getTableNames();
```
You have `getTableNames()` here whereas you have `getReferencedTables()` a few lines above. Doesn't sound very consistent.

ladislavovic replied:
`getReferencedTables()` is called on nodes and it returns tables (often it is a one table) referenced by the node. It does not returns all tables of the entity. If we rename the method to `getTableNames()`, it will not be so expressive.
`getTableNames()` is called on persiter and it just returns tables the entity is mapped to. The name is consistent with other persister methods.

gsmet commented on hibernate-core/src/main/java/org/hibernate/hql/internal/ast/tree/FromReferenceNode.java:
```diff
@@ -137,4 +137,14 @@ protected boolean isFromElementUpdateOrDeleteRoot(FromElement element) {
 
+	/**
+	 * Returns a table names which are referenced by this node. If the tables
+	 * can not be determined it returns null.
```
You say `If the tables can not be determined it returns null.` which IMHO would imply that for some reasons, we might not be able to determine the tables and in this case, we should disable the optimization. That's not what I saw in the rest of the code.

Is it just me misunderstanding or you had plans to take this into account?

If we don't have a special case for `null`, I think a collector pattern might be more efficient. It would avoid all these array initializations.

ladislavovic replied:
It does not mean we are not able to deremine tables. Some nodes just does not refer any table. Tthen it returns null, It could also returns an empty array, there is no semantic difference between them.

Long answer:  for examle expression `tableAlias.prop` is translated to following AST:
```
DotNode
  +- IdentNode (tableAlias)
  \- IdentNode (prop)
```
The second `IdentNode `has no reference to persister and also no reference to fromElement (both are null). It is impossible to get a reference to table from that. But it does not mean I miss a reference I just get it from another node. For example here I get it from DotNode. So to sum it up I get information from a subset of nodes, not from all.

gsmet commented on hibernate-core/src/main/java/org/hibernate/persister/entity/AbstractEntityPersister.java:
```diff
@@ -521,6 +521,14 @@ public boolean hasRowId() {
 	}
 
+	public String[] getTableNames() {
+		String[] tableNames = new String[ getTableSpan() ];
```
I discussed the opportunity of using a collector pattern (not sure it's the best solution, just mentioning it) but it's typically to avoid this.

Moreover, you do this work everytime you will call the method, and AFAICS the method will be called several times (e.g. for each query referencing this persister). That should be avoided if it can be.

ladislavovic replied:
I do not know collector pattern, please suggest another implementation of the method.

JWGmeligMeyling replied:
Usually it is preferred in the persisters to compute these values once, and then store them in a field. This because they are accessed many times.

For example: look at the method `doLateInit()` in `AbstractEntityPersister` which is invoked by `postInstantiate()`. I think this could be spot to instantiate the array as well.

ladislavovic commented:
I have squashed that to two commits again. I think we should finish this issue till we all have it in recent memory.

ladislavovic commented:
@Sanne  I have inherited from `ASTPrinter `in this issue. But you made `ASTPrinter ``final `a few days ago. I have removed `final `to make it compillable again. Must it be final? If so how can I customize AST tree printing? Thank you!

Sanne commented:
@ladislavovic sorry, feel free to undo my change. I was optimising a different thing, and changed it to be final just because there was no reason not to.

JWGmeligMeyling commented:
Can you add a test with classes C extends B extends A and then try to project a tuple with a property of C together with a property of A? Want to make sure the right joins are produced if B is omitted.

Or is there already a test for this?

ladislavovic commented:
> Can you add a test with classes C extends B extends A and then try to project a tuple with a property of C together with a property of A? Want to make sure the right joins are produced if B is omitted.

I added the test for that, see changes in `OmitAncestorJoinTest`.

> Same domain, but select a property from a secondarytable from B with the join to B itself being omitted.

The join to B itself will not be omitted. This issue optimize joining of superclasses tables only, it does not optimize secondary table joining. I added test for that, see `OmitAncestorJoinWhenSecondaryTablePresentTest`.

beikov commented:
Since this optimization is deactivatable in case we missed something, I think this should be safe to merge. Any objections to merge this for 5.4?

beikov commented:
@ladislavovic If you have the time, https://github.com/hibernate/hibernate-orm/pull/2482 provides a test case that uses secondary table joins that could/should be omitted with this optimization enabled. Do you want to take over that issue after we merge this? You seem to have an idea already how this could be done and it would be a really nice improvement!

gsmet commented:
Frankly, I'm not very excited about merging this to 5.4.

I really want 5.4 to be as risk free as possible so that people migrate to it as quickly as possible as it's the version that will have all the community fixes.

Note that this is particular because of the early product-freeze of 5.3.

beikov commented:
I understand that, but this optimization is deactivatable, so isn't this _risk free_ enough? When people try out 5.4 and they see something fails because of the optimization, they can easily disable it and we can get into touch with these people to fix the problems we would most probably not find anyway because we are not able to cover all possible models that are out there.

I might be wrong, but even if we communicate this optimization through the release announcements, I don't think many people will enable that optimization. On the other hand, if 5.4 doesn't work for them they get in touch and we immediatelly can help them resolve the issue, yet still get feedback about the model that we missed to cover.
I just don't see how else we could do it, even if we defer such optimizations to later releases, if the solution fails to cover cases, this might cause problems at some point anyway, so why defer it at all?
For which release would you target this PR then? Are you planning to do a 5.5 release? Considering how much the code base changed in 6.0, this PR will not work for 6.0, so this would have been wasted effort if we didn't integrate it in 5.x.

JWGmeligMeyling commented:
The problem is l that I am not even sure whether this change is spec compliant. Query outcomes will be different in some cases. A lot of other cases that are not unlikely broken by this feature remain untested. Some of my questions and issues raised still remained unanswered. This change needs to be extensively assessed imo, or disabled by default.

Rest assure, I use joined inheritance extensively and Id really like this feature to make it into Hibernate soon. And im certainly willing to contribute tests etc. But Im not looking forward to a release  require a couple hotfixes or even reverts that go unnoticed until it made it into the product.

I personally want to go to 5.4 as soon as possible as I have been running local snapshots since 5.2.12 and we need a stable release without any issues with the slightest doubt of stability.

ladislavovic commented:
>What it doesn't test is the following:
>- A secondary table associated with subclass B, with subclass C and superclass A.
>- This secondary table will be joined against the primary key of the subclass table for B.
>- This primary key will not be available, because you've just omitted the join to B, because only values from the secondary table and A are projected.

Please write the test yourself. I do not understand what exactly should the test do. Feel free to commit that into the branch, I think I let it open for project members.

JWGmeligMeyling commented:
I'll add the tests for the cases that I find important to test!

ladislavovic commented:
>I also like to see the above described test case, plus some tests using @Where and @Filter on various levels in the class hierarchy

When a filter is enabled the feature is switched off and all tables are joined. See `JoinProcessor.findQueryReferencedTables`

`@Where` could cause problems, but I try that and got interesting results. It always joined all tables in inheritance hierarchy. So probably the program goes by different way.

Moreover @Where cause error when you use it with inheritance strategy and join a subclass. And it is not related to tables omitting, it is reproducible also on master branch. The following test ends with error because generated SQL is incorrect. Generated SQL:
```
SELECT subalist1_.valsuba AS col_0_0_ 
FROM   b_table joinedinhe0_ 
       INNER JOIN suba_table subalist1_ 
               ON joinedinhe0_.id = subalist1_.b_id 
                  AND ( subalist1_.vala = 'foo' ) 
       INNER JOIN a_table subalist1_1_ 
               ON subalist1_.id = subalist1_1_.id 
```
The test:
```
package org.hibernate.test.joinwithoutancestor;

import java.util.List;
import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.Id;
import javax.persistence.Inheritance;
import javax.persistence.InheritanceType;
import javax.persistence.ManyToOne;
import javax.persistence.OneToMany;
import javax.persistence.Table;

import org.hibernate.annotations.Where;

import org.hibernate.testing.TestForIssue;
import org.hibernate.testing.junit4.BaseCoreFunctionalTestCase;
import org.hibernate.testing.transaction.TransactionUtil;
import org.junit.Test;

@TestForIssue(jiraKey = "HHH-12993")
public class JoinedInheritanceStrategyAndWhereAnnotation extends BaseCoreFunctionalTestCase {

	@Override
	protected Class[] getAnnotatedClasses() {
		return new Class[] { A.class, SubA.class, B.class };
	}

	@Test
	public void test() {
		TransactionUtil.doInHibernate( this::sessionFactory, session -> {
			String q = "select suba.valSubA from B b join b.subAList suba";
			session.createQuery( q ).getResultList();
		} );
	}

	@Entity(name = "A")
	@Table(name = A.TABLE)
	@Inheritance(strategy = InheritanceType.JOINED)
	static class A {

		public static final String TABLE = "A_Table";

		@Id
		@GeneratedValue
		private Long id;

		private String valA;

		public Long getId() {
			return id;
		}

		public String getValA() {
			return valA;
		}

		public void setValA(String valA) {
			this.valA = valA;
		}

	}

	@Entity(name = "SubA")
	@Table(name = SubA.TABLE)
	static class SubA extends A {

		public static final String TABLE = "SubA_Table";

		private String valSubA;

		@ManyToOne
		private B b;

		public String getValSubA() {
			return valSubA;
		}

		public void setValSubA(String valSubA) {
			this.valSubA = valSubA;
		}

		public B getB() {
			return b;
		}

		public void setB(B b) {
			this.b = b;
		}
	}

	@Entity(name = "B")
	@Table(name = B.TABLE)
	static class B {

		public static final String TABLE = "B_table";

		@Id
		@GeneratedValue
		private Long id;

		private String valB;

		@OneToMany(mappedBy = "b")
		@Where(clause = "valA = 'foo'")
		private List<SubA> subAList;

		public Long getId() {
			return id;
		}

		public String getValB() {
			return valB;
		}

		public void setValB(String valB) {
			this.valB = valB;
		}

		public List<SubA> getSubAList() {
			return subAList;
		}

		public void setSubAList(List<SubA> subAList) {
			this.subAList = subAList;
		}

	}

}
```

ladislavovic commented:
Guys I need to have the feature in Hibernate 5.4. It is important for my project. It can be disabled by default but I really need it there.

ladislavovic commented:
Guys, let's finish this issue somehow. It works, we do not have any test which fails due to that. If you think the risk is too high, make it disabled by default, mark it as experimental, whatever... But that would be a pity not to have this feature in Hibernate, it can speed up some kind of JPQL queries.

ladislavovic commented:
@beikov @gsmet @JWGmeligMeyling are you there?

beikov commented:
I'm sorry, but as far as I understood, the 5.4 branch won't receive any new features, only bug fixes. Correct me if I got that wrong @gsmet

ladislavovic commented:
Hi, will be the feature which is this pull reques about part of the Hibernate 6?

ladislavovic commented:
@sebersole @beikov Hi, will be the feature which is this pull reques about part of the Hibernate 6?

ladislavovic commented:
@sebersole @beikov Please can you answer my question? Thank you!

beikov commented:
Sorry, but I can't answer this question. I think this kind of optimization will most likely happen in Hibernate 6. If we have the commitment from the core team, I can work on this, as I would like to see this in 5.x as well, but without that, there is not much we can do with this PR unfortunately.

Sanne commented:
So I can't comment on specifics of this issue as I hadn't noticed this so far, but we could reasses our intention of not merging any feature improvement yet.

We were trying to avoid to create more work on the 5.x branches so that the core team could focus on ORM6, but since that was delayed a bit I guess that as long as people help by doing most of the work we could show some flexibility - that's just my suggestion though, would need @sebersole 's approval.

@beikov if you were to want to volunteer for this would it benefit ORM 6 too? It's probably not a good idea to add a feature in 5 which would then create more work to the 6 team to have feature parity when it's released. (While I guess we never committed to feature parity, it would be odd to include a new feature here and then have a new major which takes it out again)

beikov commented:
This is an optimization, not a new feature ;)

Sanne commented:
@beikov I agree :) But above you said

> I'm sorry, but as far as I understood, the 5.4 branch won't receive any new features, [...]

Whatever we call this, I just mean that I think we could be flexibile a bit more flexible.

beikov commented:
Alright, when steve is on board with this then, I'll work on it. I was of the impression that 5.x will only receive bug fixes from that point on and no further optimizations.

Sanne commented:
you're right that was the intention. But we drifted already, in fact I'm working on several other strong optimisations which are going to appear in 5.4+ versions.

sebersole commented:
If it applies cleanly then +1.  

As far as conflicts with 6.0, the only change that should have an impact is `Joinable` and that has to change already so it would be safe to just ignore that.  

The PR does change the formatting incorrectly in some of the HQLTest lines but that's minor.

sebersole commented:
Sorry for the delayed response.  I've been very "head down" on 6.0 work

sebersole commented:
I agree with what @Sanne mentions wrt 6.0 delay and getting useful things into 5.4 if possible (avoiding regressions ofc).  I went ahead, cleaned this up and applied it.

JWGmeligMeyling commented:
Unfortunately this does seem to cause a couple of regressions as pointed out in :
* https://hibernate.atlassian.net/browse/HHH-13670
* https://hibernate.atlassian.net/browse/HHH-1631?focusedCommentId=105647&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-105647

jwgmeligmeyling commented on hibernate-core/src/main/java/org/hibernate/hql/internal/ast/util/JoinProcessor.java:
```diff
@@ -94,9 +103,86 @@ public static JoinType toHibernateJoinType(int astJoinType) {
+			doFindAllNodes( node.getFirstChild(), clazz, found );
+		}
+		if ( node.getNextSibling() != null ) {
+			doFindAllNodes( node.getNextSibling(), clazz, found );
```
@beikov This is the code that caused the `StackOverflowError` for @heruan . The code seems to be fine under the assumption of a tree structure. I'm not sure in what scenario the AST could possibly lose this tree property... Does this ring a bell for you? Perhaps the query was just fairly complex and the stack relatively limited for a recursive implementation to work on the number of nodes in the tree. Perhaps then we should rewrite the implementation to do a queue based BFS instead.


```
java.lang.StackOverflowError: null
    at org.hibernate.hql.internal.ast.util.JoinProcessor.doFindAllNodes(JoinProcessor.java:113) ~[hibernate-core-5.4.6.Final.jar:5.4.6.Final]
    at org.hibernate.hql.internal.ast.util.JoinProcessor.doFindAllNodes(JoinProcessor.java:120) ~[hibernate-core-5.4.6.Final.jar:5.4.6.Final]
    at org.hibernate.hql.internal.ast.util.JoinProcessor.doFindAllNodes(JoinProcessor.java:120) ~[hibernate-core-5.4.6.Final.jar:5.4.6.Final]
    at org.hibernate.hql.internal.ast.util.JoinProcessor.doFindAllNodes(JoinProcessor.java:120) ~[hibernate-core-5.4.6.Final.jar:5.4.6.Final]
    at org.hibernate.hql.internal.ast.util.JoinProcessor.doFindAllNodes(JoinProcessor.java:120) ~[hibernate-core-5.4.6.Final.jar:5.4.6.Final]
```
